### PR TITLE
BOM-1234 --- Upgrade django-stdimage version

### PR DESCRIPTION
**Pull request by mzulqarnain1**
-Added django-dynamic-filenames dependency to replace UploadToAutoSlug
 with FilePattern
-Modified code to use FilePattern instead of deprecated method
 UploadToAutoSlug of stdimage
-Removed django-stdimage from pins
-Updated Requirements
-Updated Migrations
-Created new Migrations of ProgramType Model

### CODE REVIEW

**Review by mikix — COMMENTED** [review]
How'd you get `make upgrade` to bump stdimage? It's pinned in constraints.txt as `django-stdimage<4`

Also, most of this PR (the make upgrade bits) are already done in like, 3 other PRs right now upgrading bits of Discovery. If just one of those lands, this PR will be tiny.

**Comment by mzulqarnain1** [comment]
> How'd you get `make upgrade` to bump stdimage? It's pinned in constraints.txt as `django-stdimage<4`
> 
> Also, most of this PR (the make upgrade bits) are already done in like, 3 other PRs right now upgrading bits of Discovery. If just one of those lands, this PR will be tiny.

@mikix my bad, I forgot to add changes in constraint.txt to git. I have pushed that change too.

File: course_discovery/apps/course_metadata/models.py [hunk 1]
```diff
@@ -308,8 +308,9 @@ class ProgramType(TimeStampedModel):
                               'associated courses, but enrolled in other seat types, will NOT have their completion '
                               'of the course counted toward the completion of the program.'),
     )
+    upload_to_pattern = FilePattern(filename_pattern='media/program_types/logo_images')
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: Usually model-level fields are model fields. But this isn't? Why not just have it inline like UploadToAutoSlug was?

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -308,8 +308,9 @@ class ProgramType(TimeStampedModel):
...
     logo_image = StdImageField(
-        upload_to=UploadToAutoSlug(populate_from='name', path='media/program_types/logo_images'),
+        upload_to=upload_to_pattern,
```
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: You're dropping the `populate_from='name'` bit. Looks like you can use {instance.name} or something, like the examples in https://github.com/codingjoe/django-dynamic-filenames.

File: course_discovery/apps/course_metadata/models.py (hunk 1 above)
Comment (mzulqarnain1) [reply course_discovery/apps/course_metadata/models.py]: moved inline

File: course_discovery/apps/course_metadata/models.py (hunk 2 above)
Comment (mzulqarnain1) [reply course_discovery/apps/course_metadata/models.py]: Figuring this out.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -309,7 +309,7 @@ class ProgramType(TimeStampedModel):
                               'of the course counted toward the completion of the program.'),
     )
     logo_image = StdImageField(
-        upload_to=UploadToAutoSlug(populate_from='name', path='media/program_types/logo_images'),
+        upload_to=FilePattern(filename_pattern='media/program_types/logo_images/{instance.name:.40slug}{ext}'),
```
Comment (Ayub-Khan) [inline course_discovery/apps/course_metadata/models.py]: @mikix I think this should resolve the name issue.

File: course_discovery/apps/course_metadata/utils.py [hunk 4]
```diff
@@ -286,13 +286,14 @@ def ensure_draft_world(obj):
...
-class UploadToFieldNamePath(UploadTo):
+class UploadToFieldNamePath(FilePattern):
     """
     This is a utility to create file path for uploads based on instance field value
     """
     def __init__(self, populate_from, **kwargs):
         self.populate_from = populate_from
-        super(UploadToFieldNamePath, self).__init__(populate_from, **kwargs)
+        kwargs['populate_from'] = populate_from
+        super(UploadToFieldNamePath, self).__init__(**kwargs)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/utils.py]: Nit: might look more natural to do:
`super(UploadToFieldNamePath, self).__init__(populate_from=populate_from, **kwargs)`

File: course_discovery/apps/course_metadata/models.py (hunk 3 above)
Comment (mikix) [reply course_discovery/apps/course_metadata/models.py]: :+1:

**Review by mikix — APPROVED** [review]

File: course_discovery/apps/course_metadata/utils.py (hunk 4 above)
Comment (Ayub-Khan) [reply course_discovery/apps/course_metadata/utils.py]: @mikix i think we need to revisit our custom class. Im looking into it.

**Comment by Ayub-Khan** [comment]
@mikix review again.

File: course_discovery/apps/course_metadata/models.py (hunk 3 above)
Comment (mikix) [inline course_discovery/apps/course_metadata/models.py]: Could this be replaced with `UploadToFieldNamePath`?

File: requirements/local.txt [hunk 5]
```diff
@@ -44,8 +45,8 @@ django-rest-swagger==2.2.0
 django-simple-history==2.8.0
 django-solo==1.1.3
 django-sortedm2m==3.0.0
-django-stdimage==3.2.0
-django-storages==1.8
+django-stdimage==5.0.2
+django-storages==1.9
```
Comment (mikix) [inline requirements/local.txt]: Wait, this upgrade is bad -- it drops a boto backend we need. Disco now has a constraint on this, so re-run `make upgrade` please.

**Review by awais786 — APPROVED** [review]

**Review by jmbowman — APPROVED** [review]

**Review by mikix — APPROVED** [review]

**Pull request merged**